Fixed pretty printing and codegen error messages for empty piecewise

ramvenkat98 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

New behavior:
- Manually constructing Piecewise() now raises a type error
- Piecewise evaluated with no matching condition gives nan

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234". See
https://github.com/blog/1506-closing-issues-via-pull-requests .-->
Fixes #15005 based on the discussion there

#### Brief description of what is fixed or changed


#### Other comments


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. If there is no release notes entry for this PR,
write "NO ENTRY". The bot will check your release notes automatically to see
if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* functions
  * Piecewise with no arguments not allowed anymore
  * Piecewise evaluated with no matching conditions gives nan
<!-- END RELEASE NOTES -->

ramvenkat98 committed:
Fixed pretty printing and codegen error messages for empty piecewise

1. pprint called on an empty Piecewise function now prints "undefined"
2. codegen print functions raise the relevant "No otherwise statement error"
when encountering Piecewise functions

cbm755 commented on sympy/printing/julia.py:
```diff
@@ -418,7 +418,7 @@ def _print_yn(self, expr):
 
     def _print_Piecewise(self, expr):
-        if expr.args[-1].cond != True:
+        if len(expr.args) == 0 or expr.args[-1].cond != True:
```
I thought all of these codegen were working correctly before?

ramvenkat98 replied:
I changed this because the codegens were raising a tuple index out of range error later on in the function instead of raising the relevant ValueError that there isn't a default block.

cbm755 commented on sympy/printing/pretty/pretty.py:
```diff
@@ -1023,6 +1023,12 @@ def _print_NDimArray(self, expr):
+            D.baseline = D.height()//2
+            D.binding = prettyForm.OPEN
+            return D
+
```
This needs a test for pprint and latex

ramvenkat98 replied:
Have added the test case for both

cbm755 reviewed (changes requested):
I'm not sure if this is the fix we want but here are some suggestions to improve it.

Please check if Julia etc changes are really needed

ramvenkat98 committed:
Fixed same bug for latex printing, added test cases for pprint and latex

Fixed the same bug (reported for pprint) for latex printing too
Added test cases for pprint and latex printing as suggested by review

asmeurer commented:
I didn't know empty Piecewise was allowed. What does it represent? Is it actually used somewhere? 

I'm not a fan of "undefined" for the pretty printing. I think just printing it as `Piecewise()` is clearer.

asmeurer commented:
This could use a release notes entry.

asmeurer commented:
As noted on the issue, I think the empty `Piecewise()` should just raise an exception, or possibly give `nan`. @smichr what do you think?

ramvenkat98 committed:
Revert "Fixed same bug for latex printing, added test cases for pprint and latex"

This reverts commit ebcd67b8172030872607177b0538437410cd4c3f.

ramvenkat98 committed:
Revert "Fixed pretty printing and codegen error messages for empty piecewise"

This reverts commit 719af6f116e459fc6e32bf4643ddeedc270f04e9.

ramvenkat98 committed:
Piecewise() raises TypeError, Piecewise without matching conditions gives nan

Manually constructed Piecewise raises TypeError
Piecewise when evaluated (subs or evalf) without matching condition gives nan
A test was over-written to support this

ramvenkat98 commented:
@asmeurer @cbm755 @smichr 

Hi,

I've made changes according to the discussion in the issue thread #15005. Currently, Travis is failing to build on exactly one test and the error doesn't seem to be related to the code - could you restart it? Also, please let me know if you think this fix is ok.

Thanks

ramvenkat98 commented on sympy/functions/elementary/tests/test_piecewise.py:
```diff
@@ -84,11 +85,16 @@ def test_piecewise():
 
+    p6 = Piecewise((x, x > 0))
+    n = symbols('n', negative = True)
+    assert p6.subs(x, n) == Undefined
```
@asmeurer 

"Undefined" was defined (lol) as S.NaN in sympy/functions/elementary/tests/piecewise.py and used in the piecewise integration code, and the tests for piecewise integration - so Undefined == nan

cbm755 commented on sympy/functions/elementary/tests/test_piecewise.py:
```diff
@@ -84,11 +85,16 @@ def test_piecewise():
     assert Piecewise((1, Eq(x**2, -1)), (2, x < 0)).subs(x, I) == 1
 
+    p6 = Piecewise((x, x > 0))
+    n = symbols('n', negative = True)
```
`negative=True` (when used in kwargs)

cbm755 commented on sympy/functions/elementary/tests/test_piecewise.py:
```diff
@@ -30,9 +30,10 @@ def test_piecewise():
     # False condition is never retained
-    assert Piecewise((x, False)) == Piecewise(
-        (x, False), evaluate=False) == Piecewise()
+    raises (TypeError, lambda: Piecewise((x, False)))
```
this is a lossy change.  Add your new test but modify the existing one as:
```py
    assert Piecewise((2*x, x < 0), (x, False)) ==
           Piecewise((2*x, x < 0), (x, False), evaluate=False) ==
           Piecewise((2*x, x < 0))
```

cbm755 commented on sympy/functions/elementary/tests/test_piecewise.py:
```diff
@@ -30,9 +30,10 @@ def test_piecewise():
         Piecewise((x, Or(x < 1, x < 2)), (0, True))
     assert Piecewise((x, x < 1), (x, x < 2), (x, True)) == x
     assert Piecewise((x, True)) == x
+    #Empty Piecewise not accepted
```
`# Empty...` (with space)

cbm755 commented on sympy/functions/elementary/piecewise.py:
```diff
@@ -797,13 +797,17 @@ def _eval_subs(self, old, new):
             if c == True:
                 break
+        if not args_exist:
+            return Undefined
```
I haven't learning about `_eval_subs` but is it ok to return something other than `self.func...` here?  Is it safer to do `if not args_exist: args = ((Undefined, True))`?

cbm755 replied:
Also, is this the best place to make this change?  Is this the only code path that looks like evaluate/simplify/subs/canonicalize/etc where we want `nan` instead of exception?

This code starts with `# this is strictly not necessary,` and yet now it will be necessary.

(Note this is not a well-informed opinion: if you've thought about this and this is the right place, then say so.)

cbm755 reviewed (commented):
Note the class docs say:
```
         - If the function is evaluated at a place where all conditions are False,
          a ValueError exception will be raised.
```
which isn't true after this change (?)

ramvenkat98 committed:
Fixed issues based on reviews

Fixed issues based on the reviews given

ramvenkat98 commented:
@cbm755 

I think the modified function (_eval_subs) was already necessary. If I try to bypass this function by defining it as `def _eval_subs(self, old, new): return self.func(*self.args)`, it will not substitute values correctly. Instead, it would give for example:

```
>>> f = Piecewise((x, x>0), (x, x<-1))
>>> f.subs(x,0.5)
Piecewise((x, (x > 0) | (x < -1)))
```

I think the comment might mean that the function could have been written more compactly like so:

```
def _eval_subs(self, old, new):
        args = list(self.args)
        args_exist = False
        for i, (e, c) in enumerate(args):
            c = c._subs(old, new)
            e = e._subs(old, new)
            args[i] = (e,c)
        return self.func(*args)
```

which seems to give the correct result as expected. Instead, it is written the way it is to avoid evluating e._subs when c is known to be False and to avoid looking at the remaining (e,c) values when we've already found one c value that evaluates to True. I think this part of the function is not strictly necessarily, but makes it more efficient.

So I think this is a good place to make the change, because it is always passed through when calling the subs/evaluate methods and the function overall is already necessary.

About returning Undefined at the end, I looked through the code and I don't see an instance where returning Undefined straightaway would be a problem, but I'm not sure either so I agree that it is safer to do `if not args_exist: args = ((Undefined, True),)` there and have made that change.

Relating to what the class docs said, I'm not sure if they were true originally either because the original behavior seems to be that subs/evaluate returns `Piecewise()` when evaluated at a point where all conditions are False. In test_piecewise, I can only find one instance of a ValueError and that happens after it is rewritten as an ITE. I updated the class docs to reflect that it will instead return nan in such cases. 

I've also updated the test cases and the syntax issues pointed out.

Thanks

ramvenkat98 commented:
@asmeurer @cbm755 

I've made the changes requested in the review, are there any other changes that should be made or can this pull request be merged?

asmeurer commented on sympy/functions/elementary/tests/test_piecewise.py:
```diff
@@ -30,9 +30,13 @@ def test_piecewise():
+    assert Piecewise((2*x, x < 0), (x, False)) == \
+        Piecewise((2*x, x < 0), (x, False), evaluate=False) == \
+        Piecewise((2*x, x < 0))
+    raises (TypeError, lambda: Piecewise((x, False)))
```
I'm not sure about this one. I think it would be better for this case to also return nan, and only do TypeError for explicit `Piecewise()` with no arguments. It's possible for piecewise like these to be constructed, for instance, if a condition evaluates to False. Plus the error message would be confusing in this case.

ramvenkat98 committed:
Updated code based on review - only explicitly empty Piecewise raises error

Piecewise with arguments that are all False now return nan, not raise TypeError

ramvenkat98 commented:
@asmeurer 

I've changed the code so that only explicitly constructed empty piecewise functions raise a TypeError.

So the new behavior is as follows:

```
>>> Piecewise()
TypeError: At least one (expr, cond) pair expected.
>>> Piecewise((x, False))
nan
```

Is this ok?

Thanks

asmeurer commented:
Yes, that behavior is more reasonable to me.

ramvenkat98 commented:
@cbm755 

I think here it could still work for generators because args would be a tuple even if a generator argument was passed in (since the function argument includes *args, which would automatically package the arguments into the new tuple args). I tested this out in the following cases, where it works:
```
>>> from sympy import *; from sympy.abc import *
>>> def gen_f(): yield x; yield x > 0
... 
>>> Piecewise(gen_f())
Piecewise((x, x > 0))
>>> Piecewise(gen_f()).subs(x, 5)
5
>>> Piecewise(gen_f()).subs(x, -1)
nan
>>> def gen_f2(): yield (x, x>0); (x+1, x<=0)
... 
>>> Piecewise(*gen_f2())
Piecewise((x, x > 0), (x + 1, True))
```
I'm not sure if there is a case involving a generator that would break it, but I don't think so because I think *args always implies that args is a tuple.

ramvenkat98 commented:
@cbm755 @asmeurer 

Could this be merged? Thanks
